Objects/typeobject.c

```diff
@@ -32,8 +32,7 @@ class object "PyObject *" "&PyBaseObject_Type"
          & ((1 << MCACHE_SIZE_EXP) - 1))
 
 #define MCACHE_HASH_METHOD(type, name)                                  \
-        MCACHE_HASH((type)->tp_version_tag,                     \
-                    ((PyASCIIObject *)(name))->hash)
+    MCACHE_HASH((type)->tp_version_tag, ((Py_ssize_t)(name)) >> 3)
```

Comment:
I suspect that `id(s) >> 3` is a much worse hash function than `hash(s)`.
Is the cost of an extra memory read lower than the cost of the additional collisions?

Reply:
I think this cache is so wildly successful that it doesn't actually matter a whole lot one way or another.  It does do a little bit worse on collisions, but it doesn't seem to impact the hit rate much.  This is running python -m test.regrtest and logging the stats at the end of the run:

`>>3`:
{'total_slots': 4096, 'occupied_slots': 4096, 'num_hits': 20931459, 'num_misses': 4292, 'num_collisions': 127068, 'num_uncacheable': 1, 'num_mro_steps': 1170327}

`((PyASCIIObject *)(name))->hash`:
{'total_slots': 4096, 'occupied_slots': 4096, 'num_hits': 20922787, 'num_misses': 4304, 'num_collisions': 107251, 'num_uncacheable': 1, 'num_mro_steps': 1133300}

method_cache_hits is bumped in the hit case
method_cache_collisions is bumped when adding a new entry and replacing the old one
method_cache_misses is bumped when adding a new entry and there isn't an existing one

Reply:
Do you have any evidence as to which is faster?
If not, then the status quo wins.

My intuition is that the better spread from using the hash is likely to have better worse case performance, but in general there would be no measurable difference.

Reply:
I think per @methane running the benchmark suite it does look better with the object hash.  I've also setup a small micro-benchmark and put it in https://bugs.python.org/issue43452 and it shows a slight win there as well.  I had to write it in C to get any clear signal one way or another though, just doing the tight loop in Python seemed to have too much noise.